
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>LAMP</title>
    
    <meta name="author" content="Aika Terada">

    <!-- Le HTML5 shim, for IE6-8 support of HTML elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le styles -->
    <link href="http://a-terada.github.com/lamp/assets/themes/twitter/bootstrap/css/bootstrap.min.css" rel="stylesheet">
    <link href="http://a-terada.github.com/lamp/assets/themes/twitter/css/style.css?body=1" rel="stylesheet" type="text/css" media="all">

    <!-- Le fav and touch icons -->
  <!-- Update these with your own images
    <link rel="shortcut icon" href="images/favicon.ico">
    <link rel="apple-touch-icon" href="images/apple-touch-icon.png">
    <link rel="apple-touch-icon" sizes="72x72" href="images/apple-touch-icon-72x72.png">
    <link rel="apple-touch-icon" sizes="114x114" href="images/apple-touch-icon-114x114.png">
  -->
  </head>

  <body>

    <div class="navbar">
      <div class="navbar-inner">
        <div class="container">
          <a class="brand" href="http://a-terada.github.com/lamp">LAMP</a>
          <ul class="nav">
            
            
            


  
    
      
      	
      	<li><a href="http://a-terada.github.com/lamp/example.html">Analysis example</a></li>
      	
      
    
  
    
      
    
  
    
  
    
  
    
  
    
  



          </ul>
        </div>
      </div>
    </div>

    <div class="container">

      <div class="content">
        
<div class="page-header">
  <h1>LAMP  <small>Limitless-Arity Multiple-testing Procedure</small></h1>
</div>

<div class="row">
  <div class="span12">
    <p>LAMP is a code for multiple testing correction. Given the relationships between transcription factors (TFs) and genes and gene expression levels, the code lists up statistically significant combinations of TFs. The LAMP algorithm calibrates the Family-Wise Error Rate (FWER) at the same level of Bonferroni correction. Applying our method to your data instead of using the Bonferroni correction gives you the below advantages:</p>

<ul>
<li>Listing up significant combinations without an arity limit</li>

<li>Higher sensitivity</li>

<li>Faster calculation speed</li>
</ul>

<p>This manuscript is written for finding TF combinations. The LAMP is, however, capable of discovering other combinatorial effects, including, but not limited to, the following examples:</p>

<ul>
<li>Combinations of miRNA interactions that regulate gene expression</li>

<li>Combinations of histone acetylations and methylations that influence gene transcription</li>

<li>Combinations of single-nucleotide polymorphisms (SNPs) associates with the phenotype</li>
</ul>

<h3 id='table_of_contents'>Table of Contents</h3>
<div class='maruku_toc'><ul style='list-style: none;'><li><a href='#preparation'>Preparation</a><ul style='list-style: none;'><li><a href='#requirements'>Requirements</a></li><li><a href='#installation'>Installation</a></li></ul></li><li><a href='#demonstrations'>Demonstrations</a><ul style='list-style: none;'><li><a href='#sample_files'>Sample files</a></li><li><a href='#demonstration_1_mannwhitney_utest'>Demonstration 1: Mann-Whitney U-test</a></li><li><a href='#demonstration_2_fishers_exact_test'>Demonstration 2: Fisher&#8217;s Exact Test</a></li><li><a href='#demonstration_3_analysis_example_of_human_breast_cancer_dataset'>Demonstration 3: Analysis example of Human breast cancer dataset</a></li></ul></li><li><a href='#usage'>Usage</a></li><li><a href='#input_file_format'>Input File Format</a><ul style='list-style: none;'><li><a href='#targetfile'>Target-file</a></li><li><a href='#expressionfile'>Expression-file</a></li><li><a href='#note_for_input_files'>Note for Input Files</a></li></ul></li><li><a href='#output_format'>Output Format</a></li><li><a href='#note'>Note</a></li><li><a href='#license'>License</a></li><li><a href='#contact'>Contact</a></li></ul></div>

<h2 id='preparation'>Preparation</h2>

<h3 id='requirements'>Requirements</h3>

<p>The programs are written in python (&gt;=2.6) except frequent pattern mining. As the mining program, <a href='http://research.nii.ac.jp/~uno/codes.htm'>Linear time Closed itemset Miner (LCM) ver. 5.3</a> is used.</p>

<h3 id='installation'>Installation</h3>

<p>If you have Git, you can clone the latest version by running:</p>

<pre><code>$ git clone https://github.com/a-terada/lamp</code></pre>

<p>If you do not have Git, you can download LAMP in either <a href='https://github.com/a-terada/lamp/zipball/master'>zip</a> or <a href='https://github.com/a-terada/lamp/tarball/master'>tar</a> formats. The <code>lamp/</code> directory contains <code>lcm53.zip</code> which is the code to run frequent pattern mining. This zip file is redistributed and was downloaded from <a href='http://research.nii.ac.jp/~uno/codes.htm'>here</a>.</p>

<p>You set up the LAMP by the following command in <code>lamp/</code> directory.</p>

<pre><code>$ make</code></pre>

<h2 id='demonstrations'>Demonstrations</h2>

<p>Here, we present two examples of running LAMP by using the sample files. The details of the arguments and input and output formats are described below.</p>

<h3 id='sample_files'>Sample files</h3>

<p>There are three sample files in <code>sample/</code> directory.</p>

<ul>
<li>
<dt>sample_target.csv</dt>

<dd>The example file of target-file</dd>
</li>

<li>
<dt>sample_expression_value.tab</dt>

<dd>A sample file of expression-file. This file is used when Mann-Whitney U-test is selected to compute p-value.</dd>
</li>

<li>
<dt>sample_expression_over1.tab</dt>

<dd>A sample file of expression-file. This file is used when Fisher&#8217;s exact test is selected to compute p-value.</dd>
</li>
</ul>

<p>The dataset contains 4 TFs and 15 genes.</p>

<h3 id='demonstration_1_mannwhitney_utest'>Demonstration 1: Mann-Whitney U-test</h3>

<p>The following command finds all of the significant combinations from the files with a significance level &lt;= 0.05:</p>

<pre><code>$ python lamp.py -p u_test sample/sample_target.csv sample/sample_expression_value.tab 0.05 &gt; sample_u_test_result.txt 2&gt; sample_u_test_log.txt</code></pre>

<p>When the LAMP finishes successfully, the result is printed in <code>sample_u_test_result.txt</code>:</p>

<pre><code>--- results ---
Threshold: 0.01, Correction factor: 5 (# of target genes &gt;= 3)
# of significance: 1
Raw p-value	Adjusted p-value	Combination	# of target genes	Statistic score
0.00602414187918	0.0301207093959	TF1,TF2,TF3	5	45.5
Time (sec.): Correction factor 0.011229, P-value 0.000472, Total 0.011701</code></pre>

<p>The result indicates the following:</p>

<ul>
<li>When the raw p-value &lt;= 0.01 (calculated by 0.05/5), the combination is called significant.</li>

<li>There is one significant combination. The ternary combination of TFs TF1, TF2 and TF3 is significant. The raw p-value is 0.00602, and the adjusted p-value is 0.0301.</li>

<li>The running time is 0.0117 sec. in total.</li>
</ul>

<h3 id='demonstration_2_fishers_exact_test'>Demonstration 2: Fisher&#8217;s Exact Test</h3>

<p>The following command finds all significant combinations from the files with a significance level &lt;= 0.05:</p>

<pre><code>$ python lamp.py -p fisher sample/sample_target.csv sample/sample_expression_over1.tab 0.05 &gt; sample_fisher_result.txt 2&gt; sample_fisher_log.txt</code></pre>

<p>When the LAMP finishes successfully, the result is printed in <code>sample_fisher_result.txt</code>:</p>

<pre><code>--- results ---
Threshold: 0.01, Correction factor: 5 (# of target genes &gt;= 5)
# of significance: 1
Raw p-value	Adjusted p-value	Combination	# of target genes	Statistic score
0.00699300699301	0.034965034965	TF1,TF2,TF3	5	5.0
Time (sec.): Correction factor 0.005433, P-value 0.000172, Total 0.005605</code></pre>

<p>The result indicates the following:</p>

<ul>
<li>When the raw p-value &lt;= 0.01 (calculated by 0.05/5), the combination is called significant.</li>

<li>There is one significant combination. The ternary combination of TFs TF1, TF2 and TF3 is significant. The raw p-value is 0.00699, and the adjusted p-value is 0.0350.</li>

<li>The running time is 0.00561 sec. in total.</li>
</ul>

<h3 id='demonstration_3_analysis_example_of_human_breast_cancer_dataset'>Demonstration 3: Analysis example of Human breast cancer dataset</h3>

<p>As the example of real data analysis, we demonstrate how to prepare the input files by using gene expression profiles in human breast cancer in <a href='example.html'>Analysis example</a>.</p>

<h2 id='usage'>Usage</h2>

<p>In the LAMP directory, you can run LAMP using the following command:</p>

<pre><code>$ python lamp.py -p p-value-procedure target-file expression-file significance-level &gt; output-file</code></pre>

<p>The result is saved to the &#8216;output-file&#8217;. LAMP requires one option and three arguments:</p>

<ul>
<li>
<dl>
<dt><strong>option -p</strong></dt>

<dd>This option selects the p-value calculation procedure. &#8220;fisher&#8221; selects the Fisher&#8217;s exact test, and &#8220;u_test&#8221; selects the Mann-Whitney U-test.</dd>
</dl>
</li>

<li>
<dl>
<dt><strong>target-file</strong></dt>

<dd>The file that includes associations between TFs and genes.</dd>
</dl>
</li>

<li>
<dl>
<dt><strong>expression-file</strong></dt>

<dd>The gene expression file.</dd>
</dl>
</li>

<li>
<dl>
<dt><strong>significance-level</strong></dt>

<dd>The statistical significance threshold.</dd>
</dl>
</li>
</ul>

<p>The above command prints running progress in your console. To save the running progress to &#8216;log-file&#8217;, try using the following command:</p>

<pre><code>$ python lamp.py -p p-value-procedure target-file expression-file significance-level &gt; output-file 2&gt; log-file</code></pre>

<h2 id='input_file_format'>Input File Format</h2>

<p>LAMP receives two input files: target-file and expression-file. Their examples are in the <code>sample/</code> directory.</p>

<h3 id='targetfile'>Target-file</h3>

<p>The target-file provides the associations between TFs and their target genes. This file should be formatted according in the CSV format. <code>sample/sample_target.csv</code> is an example.</p>

<pre><code>#gene,TF1,TF2,TF3,TF4
A,1,1,1,0
B,1,1,1,0
C,1,0,0,1
D,0,0,0,0
...</code></pre>

<p>This file must start with the &#8217;#&#8217; symbol. The first line provides the names of TFs.</p>

<p>The following lines include the associations between a gene and TFs. The first column indicates the gene name. The other columns indicate whether the TF targets the gene. If the gene is targeted by the TF, the column value is &#8216;1&#8217;. If not, the value is &#8216;0&#8217;. For instance, gene A is target gene of TF1, TF2 and TF3 in the example.</p>

<p><code>sample/sampe_target.csv</code> contains 15 genes and 4 TFs. Thus, the file consists of 16 lines and 5 columns.</p>

<h3 id='expressionfile'>Expression-file</h3>

<p>The expression-file provides the gene expressions levels. The file is the tab-delimited format with two columns: the gene name and the expression level.</p>

<p>When LAMP derives the p-value using the Mann-Whitney U-test, the expressions levels are any real values. However, when LAMP calculates the p-value using Fisher&#8217;s exact test, the expression levels are 1 or 0. If the gene is up-regulated, then the second column is 1. Otherwise, it is 0.</p>

<p>The comment lines start with &#8217;#&#8217; and are ignored by the analysis.</p>

<p>As examples, we show the content of the sample files.</p>

<ul>
<li><code>samples/sample_expression_value.tab</code></li>
</ul>

<p>This file is prepared for the Mann-Whitney U-test. The second column takes real values.</p>

<pre><code>#gene	expression
A	1.5
B	1.2
C	0.9
D	0.8
...</code></pre>

<ul>
<li><code>samples/sample_expression_over1.tab</code></li>
</ul>

<p>This file is prepared for Fisher&#8217;s exact test. The values in the second column are 1 or 0.</p>

<pre><code>#gene	expression
A	1
B	1
C	0
D	0
...</code></pre>

<p>This file was generated from <code>samples/sample_expression_value.tab</code>. If the gene expression is higher than 1.0 then expression level is 1. Otherwise, it is 0.</p>

<p>Both <code>sample/sampe_expression_value.tab</code> and <code>sample/sampe_expression_over1.tab</code> describe the expressions of 15 genes.</p>

<h3 id='note_for_input_files'>Note for Input Files</h3>

<p>The gene orders in the target-file and the expression-file should be identical. For example, both <code>sample/sample_target.csv</code> and <code>sample/sample_expression_value.tab</code> consist of 15 genes from gene A to O. If the given gene sets are different or are in a different order, LAMP does not run properly.</p>

<h2 id='output_format'>Output Format</h2>

<p>The result is printed out after the &#8221;&#8212; results &#8212;&#8221; line. The first and second lines are formatted as the following:</p>

<pre><code>--- results ---
Threshold: [float], Correction factor: [integer] (# target genes &gt;= [integer])
# of significance: [integer]</code></pre>

<p><span>integer</span> and <span>float</span> are replaced with a real number in the actual result. Each value means the following:</p>

<ul>
<li>
<dt><strong>Threshold</strong></dt>

<dd>The significance threshold.</dd>
</li>

<li>
<dt><strong>Correction factor</strong></dt>

<dd>The number of testable combinations. The adjusted p-value is computed by multiplying the raw p-value by this value.</dd>
</li>

<li>
<dt><strong># of target genes</strong></dt>

<dd>The minimum number of target genes targeted by a testable TF combination.</dd>
</li>

<li>
<dt><strong># of significance</strong></dt>

<dd>The number of significant combinations.</dd>
</li>
</ul>

<p>The detected combinations are shown after the third line. Each line contains 5 columns:</p>

<ul>
<li>
<dt><strong>Raw p-value</strong></dt>

<dd>The p-value calculated using Fisher&#8217;s exact test.</dd>
</li>

<li>
<dt><strong>Adjusted p-value</strong></dt>

<dd>The adjusted p-value. The value of each combination printed by LAMP is smaller than the significance level. If the value is smaller than the inputted significance level, the combination is significant.</dd>
</li>

<li>
<dt><strong>Combination</strong></dt>

<dd>The significant combination. The TFs are delimited by &#8217;,&#8217; (comma).</dd>
</li>

<li>
<dt><strong># of target genes</strong></dt>

<dd>The number of target genes of the combination.</dd>
</li>

<li>
<dt><strong>Statistic score:</strong></dt>

<dd>The value used to compute the p-value. In case of calculation by Fisher&#8217;s exact test, the value means the number of genes that are both targeted and up-regulated. In the case of calculation by Mann-Whitney U-test, the value is the U-score.</dd>
</li>
</ul>

<p>The last line includes the calculation time (in seconds):</p>

<pre><code>Time (sec.): Correction factor [float], P-value [float], Total [float]</code></pre>

<ul>
<li>
<dt><strong>Correction factor</strong></dt>

<dd>The time to compute the correction term.</dd>
</li>

<li>
<dt><strong>P-value</strong></dt>

<dd>The time required to calculate the p-values of the testable combinations.</dd>
</li>

<li>
<dt><strong>Total</strong></dt>

<dd>Total time.</dd>
</li>
</ul>

<h2 id='note'>Note</h2>

<p>LAMP produces an LCM argument file and the LCM result files. One is saved as <code>target-file.4lcm53</code>, and the others are saved in <code>target-file.4lcm53.results/</code> directory. <!--When these files exist, LAMP reads LCM result files instead of running LCM.--></p>
<!--Thus, if the LAMP fail to finish the calculation, please command: -->
<p>If you want to remove these files, please use the following command:</p>

<pre><code>$ rm -rf target-file.4lcm53.*</code></pre>
<!--LCM related files can be removed by the command. If you omit the commands, the calculation might fail again.-->
<h2 id='license'>License</h2>

<p>LAMP is open source code issued under the BSD License.</p>

<h2 id='contact'>Contact</h2>

<p>Please contact the below address in case you find a bug, need detailed information or need application for different situation and so on.</p>

<ul>
<li>lamp_staff(AT)googlegroups.com</li>
</ul>
  </div>
</div>


      </div>

      <footer>
        <p>&copy; Aika Terada 2012 
          with help from <a href="http://jekyllbootstrap.com" target="_blank" title="The Definitive Jekyll Blogging Framework">Jekyll Bootstrap</a>
          and <a href="http://twitter.github.com/bootstrap/" target="_blank">Twitter Bootstrap</a>
        </p>
      </footer>

    </div> <!-- /container -->

    


  <script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-37265974-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>



  </body>
</html>

